service/storage_proxy.cc

```diff
@@ -5260,6 +5271,121 @@ const db::hints::host_filter& storage_proxy::get_hints_host_filter() const {
 }
 
+future<utils::UUID> storage_proxy::create_hint_queue_sync_point(const std::vector<gms::inet_address>& endpoints, clock_type::time_point deadline) {
+    return container().invoke_on(0, [endpoints, deadline] (storage_proxy& sp) mutable {
+        auto id = utils::UUID_gen::get_time_UUID();
+        sp._hint_queue_checkpoints.emplace(id);
+        // Waited indirectly by keeping a pointer to the storage_proxy.
+        // When drain_on_shutdown is triggered, hints manager will report an error
+        // and this future will resolve.
+        (void)when_all(
+            sp._hints_manager.wait_until_hints_are_replayed(endpoints, deadline),
+            sp._hints_for_views_manager.wait_until_hints_are_replayed(endpoints, deadline)
+        ).then([&sp, id, guard = sp.shared_from_this()] (std::tuple<future<>, future<>>&& f_tup) {
+            auto handle_result = [id] (future<> f, sstring kind) {
+                if (!f.failed()) {
+                    slogger.debug("Hint sync point {} for {} reached", id, kind);
+                    f.get();
+                } else {
+                    slogger.debug("An error occured when waiting for hint sync point {} for {} to resolve: {}", id, kind, f.get_exception());
+                }
+            };
+            handle_result(std::move(std::get<0>(f_tup)), "hints manager");
+            handle_result(std::move(std::get<1>(f_tup)), "hints for views manager");
+            sp._hint_queue_checkpoints.erase(id);
+        });
+
+        return make_ready_future<utils::UUID>(id);
+    });
+}
+
+future<bool> storage_proxy::check_hint_queue_sync_point(utils::UUID sync_point) {
+    return container().invoke_on(0, [sync_point] (storage_proxy& sp) {
+        return !sp._hint_queue_checkpoints.contains(sync_point);
+    });
+}
+
+future<> storage_proxy::wait_for_hints_to_be_replayed(std::vector<gms::inet_address> endpoints, seastar::abort_source& as) {
+    auto& db = _db.local();
+    const auto timeout_in_ms = db.get_config().wait_for_hint_replay_before_repair_in_ms();
+    if (timeout_in_ms <= 0) {
+        slogger.debug("wait_for_hint_replay_before_repair_in_ms is zero, so I'm not waiting for other nodes to replay their hints");
+        co_return;
+    }
+
+    const auto deadline = lowres_clock::now() + std::chrono::milliseconds(timeout_in_ms);
+
+    slogger.debug("Coordinating a request to wait until all hints are sent on {} nodes", endpoints.size());
+
+    co_await parallel_for_each(endpoints, [this, &endpoints_ = endpoints, deadline_ = deadline, &as] (gms::inet_address addr) -> future<> {
+        const auto deadline = deadline_;
+        const auto& endpoints = endpoints_;
+
+        const std::chrono::seconds retry_interval{10};
+
+        // Step 1: Create a hint queue sync point
+        utils::UUID sync_point;
+        try {
+            sync_point = co_await _messaging.send_hint_sync_point_create({ addr, 0 }, lowres_clock::now() + retry_interval, endpoints, deadline);
```

Comment:
The retry_interval here is confusing. It has nothing to do with retry. It is the timeout we want to wait for the verb at most.

Reply:
Changed, now this variable is named `wait_duration` and is used slightly differently - as a delay between sending checks.

Reply:
ok